beets/dbcore/db.py

```diff
@@ -594,20 +604,31 @@ def __iter__(self):
             # Objects are pre-sorted (i.e., by the database).
             return self._get_objects()
 
-    def _make_model(self, row):
-        # Get the flexible attributes for the object.
+    def _get_flex_attrs(self, ids):
+        # Get the flexible attributes for all ids.
         with self.db.transaction() as tx:
+            id_list = ', '.join(str(id) for id in ids)
             flex_rows = tx.query(
-                'SELECT * FROM {0} WHERE entity_id=?'.format(
-                    self.model_class._flex_table
-                ),
-                (row['id'],)
+                'SELECT * FROM {0} WHERE entity_id IN ({1})'.format(
+                    self.model_class._flex_table,
+                    id_list
```

Comment:
Inlining values into a query generally makes one nervous, although clearly SQL injection should be impossible here because we're just inserting a list of numbers. I am a little concerned we could run into weird problems, however, if the query string gets really long: when doing a `beet ls` on an entire large, library, for example. We're now asking SQLite to parse an SQL string with possibly hundreds of thousands of numbers—we may even overflow some other limit on the length of the query.

Is there possibly some other way to pass the set of IDs into SQLite? With some more engineering, for example, we could endeavor to use a join to query for the flexible attributes at the same time as the original query.

Or, perhaps an `executemany` query would be a good middle ground?

Reply:
So, did a little digging.

The default value for sqlite's sql query length limit is 1,000,000 bytes. The default value for the maximum number of host query parameters (so, the sql injection save alternative) is 999.
If my calculations are correct, this should be good for around 160000 tracks in a collection. I currently have 11,000 tracks in my collection, so I'm far away from that limit. Not sure, do you think that value is high enough or should we consider larger collections?

If we go down that path we won't avoid running more than one query. But perhaps we can do it based on a count somehow (maybe 100,000 tracks since it's a nice round number), so that in 99.9% of all use cases there'd be only one query.

Reply:
Good discovery! That is a pretty big number. It might be fine to just assume that never trips—one other question, though, is if using such a long query ever overwhelms the potential performance benefits for normal queries.

Overall, this gave me an idea for a somewhat larger refactoring here: maybe it would make sense to *semi-lazily* load the flexible attributes. That is, instead of loading them for every item as it's fetched from the database (fully lazy) or fetching them all up front (fully eager), maybe it makes sense to fetch them in "chunks," on demand. When you iterate through the result set, some reads will be slower than others because they will trigger the fetching of flexible attributes for the next "chunk." This could help make commands like `beet list` more responsive for very large queries while simultaneously solving the problem you set out to address here.